site/app/controllers/calendar/CalendarController.php

```diff
@@ -27,8 +32,179 @@ class CalendarController extends AbstractController {
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        $calendar_item = new CalendarItem();
+        try {
+            $calendar_item->setStringType($type);
+        }
+        catch (\InvalidArgumentException $e) {
+            $this->core->addErrorMessage("That is not a valid calendar item type");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+        $calendar_item->setDate(new \DateTime($date));
+        try {
+            $calendar_item->setText($text);
+        }
+        catch (\InvalidArgumentException $e) {
+            $this->core->addErrorMessage("Text exceeds 255 characters which is not allowed");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        $this->core->getCourseEntityManager()->persist($calendar_item);
+        $this->core->getCourseEntityManager()->flush();
+
+        $this->core->addSuccessMessage("Calendar item successfully added");
+        return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/calendar/items/{id}/edit", methods={"POST"})
+     * @AccessControl(role="INSTRUCTOR")
+     */
+    public function editMessage(): RedirectResponse {
+        // Checks if the values exist that are set and returns an error message if not
+        if (isset($_POST['type'])) {
+            $type = $_POST['type'];
+        }
+        else {
+            $this->core->addErrorMessage("Invalid or incorrect type given");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        if (isset($_POST['date'])) {
+            $date = $_POST['date'];
+        }
+        else {
+            $this->core->addErrorMessage("Invalid or incorrect date given");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        if (isset($_POST['text'])) {
+            $text = $_POST['text'];
+        }
+        else {
+            $this->core->addErrorMessage("Invalid or incorrect text given");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        if (isset($_POST['id'])) {
+            $id = $_POST['id'];
+        }
+        else {
+            $this->core->addErrorMessage("Invalid or incorrect id");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        if (strip_tags($text) !== $text) {
+            $this->core->addErrorMessage("HTML cannot be used in this text");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        $calendar_item = $this->core->getCourseEntityManager()->getRepository(CalendarItem::class)
+            ->findOneBy(['id' => $id]);
+
+        if ($calendar_item === null) {
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
```

Comment:
Do we want to add an error message saying why the redirect happened?  If we don't have an error message, it might confuse users who don't understand that an error condition occurred.